### PR TITLE
8331935: Add support for primitive array C1 clone intrinsic in PPC

### PR DESCRIPTION
**Pull request by varada1110**
https://bugs.openjdk.org/browse/JDK-8302850 port for PPC64

JMH Benchmark Results

```
Before :

Benchmark                 (size)  Mode  Cnt    Score   Error  Units
ArrayClone.byteArraycopy       0  avgt   15  114.107 ? 1.337  ns/op
ArrayClone.byteArraycopy      10  avgt   15  130.492 ? 0.991  ns/op
ArrayClone.byteArraycopy     100  avgt   15  139.103 ? 1.913  ns/op
ArrayClone.byteArraycopy    1000  avgt   15  321.688 ? 6.033  ns/op
ArrayClone.byteClone           0  avgt   15  227.602 ? 3.393  ns/op
ArrayClone.byteClone          10  avgt   15  237.624 ? 2.996  ns/op
ArrayClone.byteClone         100  avgt   15  239.219 ? 2.835  ns/op

ArrayClone.byteClone        1000  avgt   15  355.571 ? 2.946  ns/op
ArrayClone.intArraycopy        0  avgt   15  113.275 ? 1.099  ns/op
ArrayClone.intArraycopy       10  avgt   15  129.763 ? 1.458  ns/op
ArrayClone.intArraycopy      100  avgt   15  213.327 ? 2.524  ns/op
ArrayClone.intArraycopy     1000  avgt   15  449.650 ? 7.338  ns/op
ArrayClone.intClone            0  avgt   15  225.682 ? 3.048  ns/op
ArrayClone.intClone           10  avgt   15  234.532 ? 2.817  ns/op
ArrayClone.intClone          100  avgt   15  295.934 ? 4.925  ns/op
ArrayClone.intClone         1000  avgt   15  573.368 ? 5.739  ns/op
Finished running test 'micro:java.lang.ArrayClone'
Test report is stored in build/aix-ppc64-server-release/test-results/micro_java_lang_ArrayClone

==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   micro:java.lang.ArrayClone                            1     1     0     0   
==============================
TEST SUCCESS

Finished building target 'test' in configuration 'aix-ppc64-server-release'




After:

Benchmark                 (size)  Mode  Cnt    Score    Error  Units
ArrayClone.byteArraycopy       0  avgt   15  113.894 ?  0.993  ns/op
ArrayClone.byteArraycopy      10  avgt   15  131.455 ?  0.956  ns/op
ArrayClone.byteArraycopy     100  avgt   15  139.145 ?  3.002  ns/op
ArrayClone.byteArraycopy    1000  avgt   15  315.957 ? 14.591  ns/op
ArrayClone.byteClone           0  avgt   15   43.753 ?  3.669  ns/op
ArrayClone.byteClone          10  avgt   15   52.329 ?  1.041  ns/op
ArrayClone.byteClone         100  avgt   15  127.711 ?  3.938  ns/op

ArrayClone.byteClone        1000  avgt   15  225.937 ?  1.987  ns/op
ArrayClone.intArraycopy        0  avgt   15  113.788 ?  0.770  ns/op
ArrayClone.intArraycopy       10  avgt   15  131.980 ?  2.102  ns/op
ArrayClone.intArraycopy      100  avgt   15  213.745 ?  2.615  ns/op
ArrayClone.intArraycopy     1000  avgt   15  460.820 ?  7.106  ns/op
ArrayClone.intClone            0  avgt   15   42.074 ?  0.547  ns/op
ArrayClone.intClone           10  avgt   15   80.125 ?  1.735  ns/op
ArrayClone.intClone          100  avgt   15  207.313 ?  2.717  ns/op
ArrayClone.intClone         1000  avgt   15  326.546 ?  5.671  ns/op
Finished running test 'micro:java.lang.ArrayClone'
Test report is stored in build/aix-ppc64-server-release/test-results/micro_java_lang_ArrayClone

==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   micro:java.lang.ArrayClone                            1     1     0     0   
==============================
TEST SUCCESS

```

Hotspot compiler tests results : 

```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
>> jtreg:test/hotspot/jtreg:hotspot_compiler          1170  1168     2     0 <<
==============================
TEST FAILURE
```

2 test failures shown here is not related to code change. It is present without this changes

Reported Issue : [JDK-8331935](https://bugs.openjdk.org/browse/JDK-8331935)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331935](https://bugs.openjdk.org/browse/JDK-8331935): Add support for primitive array C1 clone intrinsic in PPC (**Enhancement** - P4)

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to [28013450](https://git.openjdk.org/jdk/pull/19250/files/280134500d4718c2ddcf294f0991667df6788aef)
 * [Amit Kumar](https://openjdk.org/census#amitkumar) (@offamitkumar - Committer) ⚠️ Review applies to [8e4e05ab](https://git.openjdk.org/jdk/pull/19250/files/8e4e05ab80bb657d86d04a055ab79601fd590732)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19250/head:pull/19250` \
`$ git checkout pull/19250`

Update a local copy of the PR: \
`$ git checkout pull/19250` \
`$ git pull https://git.openjdk.org/jdk.git pull/19250/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19250`

View PR using the GUI difftool: \
`$ git pr show -t 19250`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19250.diff">https://git.openjdk.org/jdk/pull/19250.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19250#issuecomment-2114656814)

### CODE REVIEW

**Comment by TheRealMDoerr** [comment]
The test failures will be fixed by https://github.com/openjdk/jdk/pull/19218. Unrelated.

**Review by TheRealMDoerr — COMMENTED** [review]
This looks good. Please adapt the indentation. You can mark it as ready for review.

**Review by TheRealMDoerr — CHANGES_REQUESTED** [review]
I got crashes when testing on linux ppc64le and noticed that we need one more adaptation to handle `stub == nullptr`. I suggest the following addition:
```diff
diff --git a/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp b/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
index b6d9200b261..dba662a2212 100644
--- a/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
+++ b/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
@@ -1968,7 +1968,11 @@ void LIR_Assembler::emit_arraycopy(LIR_OpArrayCopy* op) {
   int shift = shift_amount(basic_type);
 
   if (!(flags & LIR_OpArrayCopy::type_check)) {
-    __ b(cont);
+    if (stub != nullptr) {
+      __ b(cont);
+      __ bind(slow);
+      __ b(*stub->entry());
+    }
   } else {
     // We don't know the array types are compatible.
     if (basic_type != T_OBJECT) {
@@ -2089,9 +2093,9 @@ void LIR_Assembler::emit_arraycopy(LIR_OpArrayCopy* op) {
         __ add(dst_pos, tmp, dst_pos);
       }
     }
+    __ bind(slow);
+    __ b(*stub->entry());
   }
-  __ bind(slow);
-  __ b(*stub->entry());
   __ bind(cont);
 
 #ifdef ASSERT
```

**Comment by varada1110** [comment]
> I got crashes when testing on linux ppc64le and noticed that we need one more adaptation to handle `stub == nullptr`. I suggest the following addition:
> 
> ```diff
> diff --git a/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp b/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
> index b6d9200b261..dba662a2212 100644
> --- a/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
> +++ b/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
> @@ -1968,7 +1968,11 @@ void LIR_Assembler::emit_arraycopy(LIR_OpArrayCopy* op) {
>    int shift = shift_amount(basic_type);
>  
>    if (!(flags & LIR_OpArrayCopy::type_check)) {
> -    __ b(cont);
> +    if (stub != nullptr) {
> +      __ b(cont);
> +      __ bind(slow);
> +      __ b(*stub->entry());
> +    }
>    } else {
>      // We don't know the array types are compatible.
>      if (basic_type != T_OBJECT) {
> @@ -2089,9 +2093,9 @@ void LIR_Assembler::emit_arraycopy(LIR_OpArrayCopy* op) {
>          __ add(dst_pos, tmp, dst_pos);
>        }
>      }
> +    __ bind(slow);
> +    __ b(*stub->entry());
>    }
> -  __ bind(slow);
> -  __ b(*stub->entry());
>    __ bind(cont);
>  
>  #ifdef ASSERT
> ```

Hi @TheRealMDoerr , I have applied the suggested changes and I have fixed the indentation fixes. Testing is also done.
Thank you

**Review by TheRealMDoerr — APPROVED** [review]
LGTM. I'll rerun tests. Please ask somebody from your team to do a 2nd review.

**Comment by varada1110** [comment]
Thanks @TheRealMDoerr 
Hi @offamitkumar, Could you please review the code?

**Review by offamitkumar — APPROVED** [review]
Good

**Review by TheRealMDoerr — CHANGES_REQUESTED** [review]
I got test failures on AIX which need investigation: compiler/c2/Test6910605_2.java
assert(oopDesc::is_oop(s)) failed: JVM_ArrayCopy: src not an oop

**Comment by varada1110** [comment]
> I got test failures on AIX which need investigation: compiler/c2/Test6910605_2.java assert(oopDesc::is_oop(s)) failed: JVM_ArrayCopy: src not an oop

Hi @TheRealMDoerr , this test failure was not showing for me. 
I retested with 'JAVA_OPTIONS=-XX:TieredStopAtLevel=1' and the test is passing

```
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/hotspot/jtreg/compiler/c2/Test6910605_2.java
                                                         1     1     0     0   
==============================
TEST SUCCESS
```

**Comment by TheRealMDoerr** [comment]
I can reproduce it on linux with the fastdebug build.

**Comment by varada1110** [comment]
> I can reproduce it on linux with the fastdebug build.

Yes. The test failing with fastdebug build
```STDOUT:
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/home/hotspot/openjdk/jdk-varada/src/hotspot/share/prims/jvm.cpp:301), pid=27263472, tid=4884
#  assert(oopDesc::is_oop(s)) failed: JVM_ArrayCopy: src not an oop
#
# JRE version: OpenJDK Runtime Environment (23.0) (fastdebug build 23-internal-adhoc.hotspot.jdk-varada)
# Java VM: OpenJDK 64-Bit Server VM (fastdebug 23-internal-adhoc.hotspot.jdk-varada, mixed mode, emulated-client, tiered, compressed oops, compressed class ptrs, g1 gc, aix-ppc64)
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/hotspot/openjdk/jdk-varada/build/aix-ppc64-server-fastdebug/test-support/jtreg_test_hotspot_jtreg_compiler_c2_Test6910605_2_java/scratch/0/hs_err_pid27263472.log
[0.762s][warning][os] Loading hsdis library failed
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
#```

File: src/hotspot/cpu/ppc/c1_LIRGenerator_ppc.cpp [hunk 1]
```diff
@@ -885,7 +893,14 @@ void LIRGenerator::do_NewInstance(NewInstance* x) {
...
 #endif
-  CodeEmitInfo* info = state_for(x, x->state());
+  CodeEmitInfo* info = nullptr;
+  if (x->state_before() != nullptr && x->state_before()->force_reexecute()) {
+    info = state_for(x, x->state_before());
+    info->set_force_reexecute();
+  } else {
+    info = state_for(x, x->state());
+  }
+
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/ppc/c1_LIRGenerator_ppc.cpp:903]: This code seems to be integrated at the wrong place. Other platforms have it in `LIRGenerator::do_NewTypeArray`.

**Comment by TheRealMDoerr** [comment]
I also have a minor cleanup proposal for `LIR_Assembler::emit_arraycopy`:
```diff
diff --git a/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp b/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
index dba662a2212..2424d820177 100644
--- a/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
+++ b/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
@@ -1827,18 +1827,17 @@ void LIR_Assembler::emit_arraycopy(LIR_OpArrayCopy* op) {
 
   int flags = op->flags();
   ciArrayKlass* default_type = op->expected_type();
-  BasicType basic_type = default_type != nullptr ? default_type->element_type()->basic_type() : T_ILLEGAL;
+  BasicType basic_type = (default_type != nullptr) ? default_type->element_type()->basic_type() : T_ILLEGAL;
   if (basic_type == T_ARRAY) basic_type = T_OBJECT;
 
   // Set up the arraycopy stub information.
   ArrayCopyStub* stub = op->stub();
-  const int frame_resize = frame::native_abi_reg_args_size - sizeof(frame::java_abi); // C calls need larger frame.
 
   // Always do stub if no type information is available. It's ok if
   // the known type isn't loaded since the code sanity checks
   // in debug mode and the type isn't required when we know the exact type
   // also check that the type is an array type.
-  if (op->expected_type() == nullptr) {
+  if (default_type == nullptr) {
     assert(src->is_nonvolatile() && src_pos->is_nonvolatile() && dst->is_nonvolatile() && dst_pos->is_nonvolatile() &&
            length->is_nonvolatile(), "must preserve");
     address copyfunc_addr = StubRoutines::generic_arraycopy();
@@ -1873,7 +1872,7 @@ void LIR_Assembler::emit_arraycopy(LIR_OpArrayCopy* op) {
     return;
   }
 
-  assert(default_type != nullptr && default_type->is_array_klass(), "must be true at this point");
+  assert(default_type != nullptr && default_type->is_array_klass() && default_type->is_loaded(), "must be true at this point");
   Label cont, slow, copyfunc;
 
   bool simple_check_flag_set = flags & (LIR_OpArrayCopy::src_null_check |
```
Would be nice to have.

**Review by offamitkumar — CHANGES_REQUESTED** [review]
I guess you can update this as well: 
```diff
diff --git a/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp b/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
index 2424d820177..0c1e23c6353 100644
--- a/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
+++ b/src/hotspot/cpu/ppc/c1_LIRAssembler_ppc.cpp
@@ -2107,7 +2107,7 @@ void LIR_Assembler::emit_arraycopy(LIR_OpArrayCopy* op) {
     // subtype which we can't check or src is the same array as dst
     // but not necessarily exactly of type default_type.
     Label known_ok, halt;
-    metadata2reg(op->expected_type()->constant_encoding(), tmp);
+    metadata2reg(default_type->constant_encoding(), tmp);
     if (UseCompressedClassPointers) {
       // Tmp holds the default type. It currently comes uncompressed after the
       // load of a constant, so encode it.
```

**Review by TheRealMDoerr — APPROVED** [review]
Looks good and the tests have passed.

**Comment by varada1110** [comment]
Thank you @TheRealMDoerr @offamitkumar . I am running the tests: hotspot_compiler, hotspot_gc, hotspot_serviceability and hotspot_runtime for tier1, tier2  and tier3 with fastdebug, slowdebug and release. I will update the results.

**Comment by offamitkumar** [comment]
> with fastdebug, slowdebug and release.

I think with fastdebug is sufficient.

**Comment by shipilev** [comment]
There is currently a regression in the original code, [JDK-8332670](https://bugs.openjdk.org/browse/JDK-8332670), which may explain some instability on PPC.

**Comment by varada1110** [comment]
> I've put it again into our nightly tests and haven't seen any errors which may have been caused by this PR. There are currently some unrelated errors. So, I think it's good to go.

**Comment by varada1110** [comment]
> Thank you @TheRealMDoerr @offamitkumar . I am running the tests: hotspot_compiler, hotspot_gc, hotspot_serviceability and hotspot_runtime for tier1, tier2 and tier3 with fastdebug, slowdebug and release. I will update the results.

Completed the testing for fastdebug. There are few unrelated test failures

**Comment by TheRealMDoerr** [comment]
https://github.com/openjdk/jdk/pull/19538 is integrated, so we can ship this one, too.

**Pull request closed without merging**